src/or/entrynodes.c

```diff
@@ -1343,12 +1357,51 @@ sampled_guards_update_from_consensus(guard_selection_t *gs)
     }
 
     if (rmv) {
-      ++n_changes;
+      (*n_changes)++;
       SMARTLIST_DEL_CURRENT(gs->sampled_entry_guards, guard);
       remove_guard_from_confirmed_and_primary_lists(gs, guard);
       entry_guard_free(guard);
     }
   } SMARTLIST_FOREACH_END(guard);
+}
+
+/**
+ * Update the status of all sampled guards based on the arrival of a
+ * new consensus networkstatus document.  This will include marking
+ * some guards as listed or unlisted, and removing expired guards. */
+STATIC void
+sampled_guards_update_from_consensus(guard_selection_t *gs)
+{
+  tor_assert(gs);
+
+  // It's important to use only a live consensus here; we don't want to
+  // make changes based on anything expired or old.
+  if (live_consensus_is_missing(gs)) {
+    log_info(LD_GUARD, "Not updating the sample guard set; we have "
+             "no live consensus.");
+    return;
+  }
+  log_info(LD_GUARD, "Updating sampled guard status based on received "
+           "consensus.");
+
+  int n_changes = 0;
+
+  /* First: Update listed/unlisted. */
+  sampled_guards_update_listedness(gs, &n_changes);
+
+  const time_t remove_if_unlisted_since =
```

Comment:
Perhaps we can do these computations inside `sampled_guards_prune_obsolete_entries()`? Or you like them here?

Reply:
Sounds good!